### ISSUE TITLE
HLS plugin is failing on multiple streams

### ISSUE THREAD

**Issue by sou7611** [streamlink/streamlink#1231]
Seeing read timeout errors, forcing streamlink to exit across multiples of HLS sources.

Been able to verify on foxsportsgo, watchespn, nbcsports, mls-live, and many others.

Run flawlessly for a period of time, then abruptly exits with "Read timeout; exiting"

There is no network interruption at all, nothing on the PC, nothing to interrupt the stream.

This happens with latest 0.8.1 version.

**Comment by gravyboat** [comment]
We're going to need more details before we can provide any support on this (as noted in the template), can you rerun the commands with verbose output and debug logging please? Add  `-v` and debug `-l debug` to your command.

**Comment by back-to** [comment]
@sou7611 
you could try to use `--hls-timeout` and all the other commands https://github.com/streamlink/streamlink/issues/1146#issuecomment-320471910
https://streamlink.github.io/cli.html#cmdoption-hls-timeout

but you need to provide **more details**

- do you give streamlink all the required cookies?
- does your url expire after some seconds?

**Comment by sou7611** [comment]
Still seeing this even with latest version.

Yes all required cookies present, the URL does not expire.

You can start multiple copies and watch in MPC-HC and the stream keeps going, while it times out and stops dead while recording.

Doesn't matter the source, happens on EVERY SINGLE hls source ... espn, nbc, you name it, this happens on all of them.

Streamlink is not correctly restarting when whatever happens, happens, no mater what you do with hls-timeout or any of the other commands.

**Comment by sou7611** [comment]
For example, saw this multiple times tonight, on NBC ... which on Comcast/Xfinity doesn't even require a cookie (it authenticates from your IP) this is adding every flag I can think of to try and buffer it through whatever's causing the problem.  Over the course of a 4 hour football game, had a dozen + read timeouts.

Could plainly see other windows playing the same exact stream not affected, so it wasn't network or source related.  Was clearly streamlink failing.

Streamlink.exe  --stream-segment-attempts 500 --stream-segment-timeout 3 --player-continuous-http --hls-playlist-reload-attempts=10 --hls-live-edge=1 --hls-segment-attempts=10  --http-header "User-Agent='Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'" "hlsvariant://http://sprtlive6.akamaized.net/hls/live/586267-b/nbc-sports0113120009-ua/OTT06_VIDEO_0_5564000.m3u8" best

**Comment by back-to** [comment]
Remove `--stream-segment-attempts 500 --stream-segment-timeout 3` it won't change anything for a hlsstream

you can use `--hls-segment-timeout` https://streamlink.github.io/cli.html#cmdoption-hls-segment-timeout

`--hls-timeout` should be used, update it to 10 minutes https://streamlink.github.io/cli.html#cmdoption-hls-timeout

did you try `--hls-live-edge` with the default 3? https://streamlink.github.io/cli.html#cmdoption-hls-live-edge

sidenote `hlsvariant://` is now `hls://`

a `-l debug` for the last part of the stream when it fails would be interesting https://streamlink.github.io/cli.html#cmdoption-l

```
Streamlink.exe --player-continuous-http \
--hls-playlist-reload-attempts=10 \
--hls-live-edge=3 \
--hls-segment-attempts=10 \
--http-header "User-Agent='Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'" \
--hls-timeout 600 \
-l debug \
"hls://URL" best
```

**Comment by aedicted** [comment]
I am having the exact same issue with the latest Streamlink v0.14.2 under Windows 10 in conjuction with ORF streams such as:

http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8

The only "authentication" in place is the geo blocking to Austrian IP-addresses, other than that, no login or any cookie sessions required, so that can't be the problem.

Streams are getting downloaded perfectly fine for one or several hours but at some point, Streamlink gives the beloved "read timeout" error message and exits for no obvious reason. Already tried all the timeout options for HTTP, HLS, etc., doesn't change anything which is frustrating.

**Comment by back-to** [comment]
@aedicted 
for your example ORF,

when a stream switches from **not blocked** to **geo blocked**

ORF replaces the whole stream with a geo blocked video,
which won't play in streamlink because it's data number is different than the stream before so it's invalid for streamlink.

they might do the same for other reasons,
such as TV shows that they are not allowed to air on the internet.

`error: Error when reading from stream: Read timeout, exiting`

Here are some examples why this happens,
it will happen when a playlist
- gets replaced with a new stream which has different numbers
- expires after time
- needs cookies from the master playlist
- needs a ping back from the website or master playlist

---

**ORF2 example geo block**

**Valid playlist**

```
[stream.hls][debug] MEDIA-SEQUENCE: 4279233
[stream.hls][debug] Adding segment 4279235 to queue
[stream.hls][debug] Download of segment 4279235 complete
[stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:4279234
#EXTINF:10.000,
manifest_2_4279234.ts?m=
#EXTINF:10.000,
manifest_2_4279235.ts?m=
#EXTINF:10.000,
manifest_2_4279236.ts?m=
```

**Invalid playlist: TV show is now geo blocked**

```
[stream.hls][debug] MEDIA-SEQUENCE: 4279234
[stream.hls][debug] Adding segment 4279236 to queue
[stream.hls][debug] Download of segment 4279236 complete
[stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-ALLOW-CACHE:NO
#EXT-X-VERSION:3
#EXT-X-STREAM-INF:BANDWIDTH=2285887,CODECS="avc1.77.31,mp4a.40.2",RESOLUTION=960x544
http://apasfiis.sf.apa.at/ipad/gp/geoprotection_Q6A.mp4/chunklist.m3u8-...
[stream.hls][warning] Failed to reload playlist: Attempted to play a variant playlist, use '...' instead
[stream.hls][debug] Reloading playlist

...

error: Error when reading from stream: Read timeout, exiting
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

**Comment by aedicted** [comment]
@back-to

Thanks for your elaborate reply!

However, I don't think that any geo blocking switch is the reason for those interruptions in that particular case because they even happen occasionally during the same show like ballet today. In the middle of the performance, Streamlink exits and that will be hardly due to a spontaneous "oh, lets geo block the close up of our dancers right now because it's fun". ;)

**Comment by back-to** [comment]
@aedicted 
if you can easily reproduce it

replace the file `src/streamlink/stream/hls.py`

with
https://gist.githubusercontent.com/back-to/94fcdab116f9eb6c809d088950e8e1a5/raw/1e059a404b8e8200900cc3d8afc4eb4584f6a843/hls.py

there is only one line changed
https://gist.github.com/back-to/94fcdab116f9eb6c809d088950e8e1a5#file-hls-py-L189

this will output the used html text like this in your terminal,

I'm geo blocked so I can't test it for your case.

---

this data would help to sort your problem,
as it is a lot of spam only the data **when the playlist switches / timeouts** is interesting

also if you want to post the data here,
remove your personal IP from the playlist as it will be in the playlist URL

```
$ streamlink http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8 -l info
[cli][info] Found matching plugin hls for URL http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8
[cli][info] Available streams: 544p (worst, best)
[cli][info] Opening stream: 544p (hls)
#EXTM3U
#EXT-X-ALLOW-CACHE:NO
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:11
#EXT-X-MEDIA-SEQUENCE:0
#EXTINF:10.031,
http://varorfvod.sf.apa.at/system_clips/geoprotection_Q6A.mp4/media_0.ts?...
#EXT-X-ENDLIST


[cli][info] Starting player: /usr/bin/mpv
#EXTM3U
#EXT-X-ALLOW-CACHE:NO
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:11
#EXT-X-MEDIA-SEQUENCE:0
#EXTINF:10.031,
http://varorfvod.sf.apa.at/system_clips/geoprotection_Q6A.mp4/media_0.ts?...
#EXT-X-ENDLIST


#EXTM3U
#EXT-X-ALLOW-CACHE:NO
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:11
#EXT-X-MEDIA-SEQUENCE:0
#EXTINF:10.031,
http://varorfvod.sf.apa.at/system_clips/geoprotection_Q6A.mp4/media_0.ts?...
#EXT-X-ENDLIST
```

**Comment by aedicted** [comment]
Hi back-to, thanks for your support here, I really appreciate it.

I will try to follow your steps and test it. Yeah, it is reproducable however takes several hours for the small 720p streams to occur. It seems to occur more often with the 4K/HDR stream which blasts at about 40 Mbps, though. Maybe I take that one, then. :)

By the way, in order to further narrow it down and to support my argument that it can't be due to any geo block switching, I recorded the same stream now in several instances and et voila, Streamlink exits at different positions, so when one recording ends at point x, the other one passes that point and ends at point y. Hence it can't be the source, otherwise both would quit at exactly the same point.

**Comment by aedicted** [comment]
So, I've changed the file which for me resides under C:\Program Files (x86)\Streamlink\pkgs\streamlink\stream and that gives me the following output while the stream is running:

C:\Program Files (x86)\Streamlink\bin>streamlink.exe http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8 best
[cli][info] Found matching plugin hls for URL http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8
[cli][info] Available streams: 720p (worst, best)
[cli][info] Opening stream: 720p (hls)
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:130329
#EXTINF:10.0,
media-ugbja9stn_130329.ts
#EXTINF:10.0,
media-ugbja9stn_130330.ts
#EXTINF:10.0,
media-ugbja9stn_130331.ts

[cli][info] Starting player: "C:\Program Files\VideoLAN\VLC\vlc.exe"
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:130331
#EXTINF:10.0,
media-ugbja9stn_130331.ts
#EXTINF:10.0,
media-ugbja9stn_130332.ts
#EXTINF:10.0,
media-ugbja9stn_130333.ts

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:130332
#EXTINF:10.0,
media-ugbja9stn_130332.ts
#EXTINF:10.0,
media-ugbja9stn_130333.ts
#EXTINF:10.0,
media-ugbja9stn_130334.ts

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:130334
#EXTINF:10.0,
media-ugbja9stn_130334.ts
#EXTINF:10.0,
media-ugbja9stn_130335.ts
#EXTINF:10.0,
media-ugbja9stn_130336.ts

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:130335
#EXTINF:10.0,
media-ugbja9stn_130335.ts
#EXTINF:10.0,
media-ugbja9stn_130336.ts
#EXTINF:10.0,
media-ugbja9stn_130337.ts

**Comment by aedicted** [comment]
I tested further and with the UHD-stream it is the same, one instance exits, the other keeps recording. However, while most of the MEDIA-SEQUENCES are continous, once it a while, there is one missing, like this:

[download][orf_uhd_test.ts] Written 18.14 GB (1h9m25s @ 6.4 MB/s)
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:736145
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736145.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736146.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736147.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736148.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736149.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736150.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736151.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736152.ts?m=1530090172&f=5

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:736147
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736147.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736148.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736149.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736150.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736151.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736152.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736153.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_736154.ts?m=1530090172&f=5

So I wonder if data gets lost or not because there is no error message about that a sequence couldn't be downloaded. Also, the parts missing are the same in all instances, in this case 73614**6**.

**Comment by beardypig** [comment]
@aedicted the media sequence is just the number of the first media item in the list, there can be gaps in this number. However, it must continue to increase, it can never decrease and if it stays the same for too long (a few reloads) then the stream will timeout.

**Comment by aedicted** [comment]
@beardypig many thanks to you as well. Okay, then it should be okay and all segments be present as long as there is no error message, right?

While continuing the testing further, in some cases, Streamlink complains about segments it couldn't retrieve, however chronologically, way **after** the current sequence, like dozens of segments later. Looking at the ouput, one would assume that those segments have been written to the file already long ago.

Furthermore, also these missed segments then were retrieved without any error message in the other Streamlink instance running in parallel, so something odd is going on within a given instance. In other words: if the other Streamlink instance which runs in parallel technically was able to retrieve the segments in question in time, then the download stragety at some point must be sub-optimal or buggy.

**Comment by beardypig** [comment]
Yeah, there are definitely some improvements to be made. I’ve been working on it for a while - but it complex and I have less time than I’d like to work on it :)

**Comment by aedicted** [comment]
Unfortunately I am not a programmer but I'd be the first to help you testing any update you might be able to create.

Today, my little project was to record the 4K transmission from the austrian TV station ORF of the game "Belgium - England".

As expected, Streamlink also there exited randomly so I did this with two instances with a for loop in order to cover the file naming (1, 2, and so on) and to lose as little as possible per instance. This attempt was raising the bitrate to 80-100 Mbps which is the opposite of efficiency but luckily doable with both my internet access and VPN provider.

For the 4K stream, the ORF transmits it's logo as a still frame but full bitrate video so one can test the stream even before the game starts. I started recording about an hour before the transmission even began and interestingly, at first, both instances had an interuption about 5 GB, exactly the same file size. This was not the moment when the transmission began (that was 20 minutes after the first interruption) so it is up to speculation what caused Streamlink to stop at the same point here.

After that, there were two interruptions in both instances, but again luckily at two totally different positions so I was able to grab some Hex Editor and sync the whole thing manually to have one complete stream (about 20 GB I could easily compare are binary identical so I assume that at least large parts are completely free of errors).

So yeah, one can create several instances as a workaround, but that causes unnecessary traffic and work of course so it would be good to investigate why Streamlink stops recording in the first place.

**Comment by aedicted** [comment]
Sure, luckily I redirected the output to text files.

So after the first interruption which affected both instances at exactly the same point (no program content at this point so no real loss at least):

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:742730
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742730.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742731.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742732.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742733.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742734.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742735.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742736.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_742737.ts?m=1530090172&f=5

error: Error when reading from stream: Read timeout, exiting
[cli][info] Stream ended

After that since it were different points, **first instance**:

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:743688
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743688.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743689.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743690.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743691.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743692.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743693.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743694.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743695.ts?m=1530090172&f=5

error: Error when reading from stream: Read timeout, exiting
[cli][info] Stream ended
[cli][info] Closing currently open stream...

And the **second one**

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:743168
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743168.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743169.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743170.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743171.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743172.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743173.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743174.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_743175.ts?m=1530090172&f=5

error: Error when reading from stream: Read timeout, exiting
[cli][info] Stream ended


Also this time, the logs don't contain any error message in terms of unretrievable parts due to HTTP errors.

**Comment by back-to** [comment]
@aedicted 

> So yeah, one can create several instances as a workaround, but that causes unnecessary traffic 

I have something to reload invalid playlists so you don't need this,
but it depends on the error to have it work correctly that is why I want some more details.

> Also this time, the logs don't contain any error message in terms of unretrievable parts due to HTTP errors.

could you post this error

---

well just one html text is not enough,
I wanted to see the changes and what streamlink is doing with it

you can use https://gist.github.com/ to upload longer logs

```
[debug] Download of segment
[debug] Adding segment
```

I would like something like this with ~3 successful `Download of segment X complete` messages

```
[20:07:27,883][stream.hls][debug] Download of segment 128698 complete
[20:07:27,935][stream.hls][debug] Download of segment 128699 complete
[20:07:29,38][stream.hls][debug] Reloading playlist
[20:07:29,367][stream.hls][debug] Adding segment 128700 to queue
[20:07:43,376][stream.hls][debug] Reloading playlist
[20:07:44,531][stream.hls][debug] Adding segment 128701 to queue
[20:07:44,531][stream.hls][debug] Adding segment 128702 to queue
[20:07:58,534][stream.hls][debug] Reloading playlist
[20:07:59,63][stream.hls][debug] Adding segment 128703 to queue
[20:08:13,73][stream.hls][debug] Reloading playlist
[20:08:14,38][stream.hls][debug] Adding segment 128704 to queue
[20:08:14,38][stream.hls][debug] Adding segment 128705 to queue
[20:08:28,40][stream.hls][debug] Reloading playlist
error: Error when reading from stream: Read timeout, exiting
[20:08:28,45][stream.segmented][debug] Closing worker thread
[20:08:28,46][stream.segmented][debug] Closing writer thread
[20:08:28,47][cli][info] Stream ended
[20:08:28,47][cli][info] Closing currently open stream...
```

**Comment by aedicted** [comment]
Do you need output like that?:

C:\Program Files (x86)\Streamlink\bin>streamlink.exe http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8 best --hls-segment-threads 10 --hls-segment-attempts 10 --hls-playlist-reload-attempts 10 --hls-segment-timeout 10 --http-stream-timeout 10 -l trace -o d:\temp\debug.ts
[05:13:55,222][cli][debug] OS:         Windows 10
[05:13:55,224][cli][debug] Python:     3.5.2
[05:13:55,225][cli][debug] Streamlink: 0.14.2
[05:13:55,225][cli][debug] Requests(2.19.1), Socks(1.6.7), Websocket(0.48.0)
[05:13:55,225][cli][info] Found matching plugin hls for URL http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8
[05:13:55,226][plugin.hls][debug] URL=http://orfs.orf.cdn.ors.at/orfs/q8c.sdp/playlist.m3u8; params={}
[05:13:59,414][cli][info] Available streams: 720p (worst, best)
[05:13:59,415][cli][info] Opening stream: 720p (hls)
[05:13:59,417][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:140617
#EXTINF:10.0,
media-ugbja9stn_140617.ts
#EXTINF:10.0,
media-ugbja9stn_140618.ts
#EXTINF:10.0,
media-ugbja9stn_140619.ts

[05:13:59,743][stream.hls][debug] First Sequence: 140617; Last Sequence: 140619
[05:13:59,743][stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 140617; End Sequence: None
[05:13:59,745][stream.hls][debug] Adding segment 140617 to queue
[05:13:59,745][cli][debug] Pre-buffering 8192 bytes
[05:13:59,746][stream.hls][debug] Adding segment 140618 to queue
[05:13:59,749][stream.hls][debug] Adding segment 140619 to queue
[05:14:13,755][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:140619
#EXTINF:10.0,
media-ugbja9stn_140619.ts
#EXTINF:10.0,
media-ugbja9stn_140620.ts
#EXTINF:10.0,
media-ugbja9stn_140621.ts

[05:14:14,750][stream.hls][debug] Adding segment 140620 to queue
[05:14:14,753][stream.hls][debug] Adding segment 140621 to queue
Interrupted! Exiting...
[05:14:19,594][cli][info] Closing currently open stream...
[05:14:19,601][stream.hls][debug] Download of segment 140617 complete


If so, I'll keep it running through the night here and send you the result in a few hours.

**Comment by back-to** [comment]
@aedicted 

I just need the last 4 or 5 with `Adding, Download, Reloading playlist ...`

before the error, if you still have it in your old logs it could be enough

I want to see if the playlist changes or if it is a cached / same playlist

---

**I don't know if this will work with your error**,
but if you want to test something experimental

create this folder

`%APPDATA%\streamlink\plugins`

https://streamlink.github.io/cli.html#sideloading-plugins

download this file `hlssession.py` and put it in there

https://raw.githubusercontent.com/back-to/plugins/master/plugins/hlssession.py

run

```
streamlink --hlssession-segment --hlssession-ignore-number 20 hlssession://http://...YOUR URL
```

this will reload your session m3u8 URL when the playlist doesn't change after two reloads

you can also add `--hlssession-time 01:00:00` which will reload your playlist URL after an hour automatically

but like I said I don't know if this works with your error
it is originally meant for playlist which expires

**Comment by aedicted** [comment]
Since for yesterday's recording I only have the normal output and not the debug/trace one, I give it another run and post the result here soon.

Meanwhile I tried to test the plugin you suggested, however I get the following error message then:

C:\Program Files (x86)\Streamlink\bin>streamlink --hlssession-segment --hlssession-ignore-number 20 hlssession://http://ors-uhd.cdn.ors.at/out/u/orfuhd-abr-hdr10_1.m3u8
Failed to load plugin hlssession:
  File "imp.py", line 234, in load_module
  File "imp.py", line 172, in load_source
  File "<frozen importlib._bootstrap>", line 693, in _load
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "C:\Users\aedicted\AppData\Roaming\streamlink\plugins\hlssession.py", line 13, in <module>
    from streamlink.utils.args import num
ImportError: No module named 'streamlink.utils.args'

usage: streamlink [OPTIONS] <URL> [STREAM]
streamlink: error: unrecognized arguments: --hlssession-segment --hlssession-ignore-number

**Comment by aedicted** [comment]
Here the log file in trace mode with the ORF Sport HD stream:

[18:55:30,135][stream.hls][debug] Adding segment 145548 to queue
[18:55:44,137][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145548
#EXTINF:10.0,
media-ugbja9stn_145548.ts
#EXTINF:10.0,
media-ugbja9stn_145549.ts
#EXTINF:10.0,
media-ugbja9stn_145550.ts

[18:55:45,98][stream.hls][debug] Adding segment 145549 to queue
[18:55:45,98][stream.hls][debug] Adding segment 145550 to queue
[18:55:56,166][stream.hls][debug] Download of segment 145548 complete
[18:55:56,177][stream.hls][debug] Download of segment 145549 complete
[18:55:56,190][stream.hls][debug] Download of segment 145550 complete
[18:55:59,99][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145549
#EXTINF:10.0,
media-ugbja9stn_145549.ts
#EXTINF:10.0,
media-ugbja9stn_145550.ts
#EXTINF:10.0,
media-ugbja9stn_145551.ts

[18:55:59,414][stream.hls][debug] Adding segment 145551 to queue
[18:56:07,159][stream.hls][debug] Download of segment 145551 complete
[18:56:13,416][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145551
#EXTINF:10.0,
media-ugbja9stn_145551.ts
#EXTINF:10.0,
media-ugbja9stn_145552.ts
#EXTINF:10.0,
media-ugbja9stn_145553.ts

[18:56:13,749][stream.hls][debug] Adding segment 145552 to queue
[18:56:13,749][stream.hls][debug] Adding segment 145553 to queue
[18:56:27,750][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145552
#EXTINF:10.0,
media-ugbja9stn_145552.ts
#EXTINF:10.0,
media-ugbja9stn_145553.ts
#EXTINF:10.0,
media-ugbja9stn_145554.ts

[18:56:28,119][stream.hls][debug] Adding segment 145554 to queue
[18:56:42,120][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145554
#EXTINF:10.0,
media-ugbja9stn_145554.ts
#EXTINF:10.0,
media-ugbja9stn_145555.ts
#EXTINF:10.0,
media-ugbja9stn_145556.ts

[18:56:42,485][stream.hls][debug] Adding segment 145555 to queue
[18:56:42,485][stream.hls][debug] Adding segment 145556 to queue
[18:56:56,486][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:14
#EXT-X-MEDIA-SEQUENCE:145555
#EXTINF:10.0,
media-ugbja9stn_145555.ts
#EXTINF:10.0,
media-ugbja9stn_145556.ts
#EXTINF:10.0,
media-ugbja9stn_145557.ts

[18:56:56,812][stream.hls][debug] Adding segment 145557 to queue
error: Error when reading from stream: Read timeout, exiting
[18:57:07,186][stream.segmented][debug] Closing worker thread
[18:57:07,187][stream.segmented][debug] Closing writer thread
[18:57:07,188][cli][info] Stream ended
[18:57:07,190][cli][info] Closing currently open stream...

**Comment by aedicted** [comment]
And here from one instance of the UHD stream:

19:15:57,399][stream.hls][debug] Adding segment 750589 to queue
[19:16:00,528][stream.hls][debug] Download of segment 750587 complete
[19:16:07,401][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:750583
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750583.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750584.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750585.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750586.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750587.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750588.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750589.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750590.ts?m=1530090172&f=5

[19:16:07,734][stream.hls][debug] Adding segment 750590 to queue
[19:16:17,738][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:750585
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750585.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750586.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750587.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750588.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750589.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750590.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750591.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750592.ts?m=1530090172&f=5

[19:16:21,891][stream.hls][debug] Adding segment 750591 to queue
[19:16:21,892][stream.hls][debug] Adding segment 750592 to queue
[19:16:31,895][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:750586
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750586.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750587.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750588.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750589.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750590.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750591.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750592.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750593.ts?m=1530090172&f=5

[19:16:32,854][stream.hls][debug] Adding segment 750593 to queue
[19:16:42,855][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:750587
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750587.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750588.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750589.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750590.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750591.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750592.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750593.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750594.ts?m=1530090172&f=5

[19:16:43,983][stream.hls][debug] Adding segment 750594 to queue
[19:16:53,984][stream.hls][debug] Reloading playlist
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:750588
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750588.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750589.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750590.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750591.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750592.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750593.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750594.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_750595.ts?m=1530090172&f=5

[19:16:55,863][stream.hls][debug] Adding segment 750595 to queue
error: Error when reading from stream: Read timeout, exiting
[19:17:00,620][stream.segmented][debug] Closing worker thread
[19:17:00,621][stream.segmented][debug] Closing writer thread
[19:17:00,622][cli][info] Stream ended
[19:17:00,623][cli][info] Closing currently open stream...

**Comment by aedicted** [comment]
And here the result of yesterday's final stream (Croatia vs. France):

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:752700
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752700.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752701.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752702.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752703.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752704.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752705.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752706.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752707.ts?m=1530090172&f=5

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:752702
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752702.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752703.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752704.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752705.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752706.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752707.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752708.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752709.ts?m=1530090172&f=5

error: Error when reading from stream: Read timeout, exiting
[cli][info] Stream ended
[cli][info] Closing currently open stream...
[cli][info] Found matching plugin hls for URL http://at-ors-o3.yaremedia.net/out/u/orfuhd-abr-hdr10_1.m3u8
[cli][info] Available streams: live (worst, best)
[cli][info] Opening stream: live (hls)
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:752702
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752702.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752703.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752704.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752705.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752706.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752707.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752708.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752709.ts?m=1530090172&f=5

#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:752703
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752703.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752704.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752705.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752706.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752707.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752708.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752709.ts?m=1530090172&f=5
#EXTINF:10.000,
orfuhd-abr-hdr10_1_752710.ts?m=1530090172&f=5

Although the log here (for me at least) suggests differently, the data gap in between was 221 MB (and another one 175 MB). Luckily, both I was able to recover from the second stream recording running.

Total airing was 71 GB for about 4.7 hours - nice to have that archived so thanks to have that tool developed. :)

**Comment by beardypig** [comment]
@aedicted how large is each segment? Must be pretty big, 40-45MB each? 

The error you are getting is when the download buffer is empty, it only waits a certain amount of time (60 seconds) for there to be more data - I'd recommend trying with fewer threads and see if that helps. 

This settings is controlled by `--stream-timeout` which has a default value of 60 seconds. If you are downloading 10 segments at a time, 45MB each, 450MB ... that _should_ finish within 60 seconds - assuming your connection is fast enough. If it goes slow for one segment you might have problems, or if a download stalls you _will_ have problems. I think this might actually be the same problem as #851 (I'm not sure @back-to was absolutely correct with the timeouts because it could just go slow or be stalled). You could try setting `--stream-timeout 600` and see if that allows you to download the stream. 

I'm working on a new mechanism for downloading segmented streams that will improve this and would make it easier to retry stalled downloads.

**Comment by aedicted** [comment]
@beardypig 

Yes, you are right. One segment is about 44MB. If I use wget and download a single one, the download speed often only is about 2-3 MB/s at the end starting with a few hundred kB/s.

So throughput-wise, the connection is fast enough to transmit those 450 MB within 60 seconds but due to TCP I greatly assume that some single segment downloads will be too slow.

I will try again in four years at latest haha, but maybe some other demanding UHD content will show up soon, let's hope so. :)

But don't forget that those random dropouts also continue with regular, way less demanding HD streams with a nominal bitrate of 3~4 Mbps only.

Also do you know how long those valids are valid with HLS? Is that depending on the server config?

**Comment by hmedia1** [comment]
I've been using streamlink for years since the days of livestreamer for both YouTube and also hls streams from various cloud servers direct.

I've been seeing this error come up in a very inconsistent way throughout many versions and across many different environments ranging from low-quality 3G/4G internet, right through to on-premise uncontended fibre internet.

Similarly I've seen the low quality lines pull entire streams and the high quality ones failing, and vice versa ... using `cURL` to download segments to debug is always simple and never times out.  It's like it just goes from bulletproof, to completely failed in any random instant without reason.

For example, one stream today;

<pre><code>
 streamlink --hls-segment-threads 2 \
--hls-segment-attempts 1000 \
--hls-segment-timeout 60 \
--hls-timeout 60 \
--http-stream-timeout 60 \
--retry-open 9999 \
--retry-streams 9999 \
--hls-live-restart \
"http://server/hls/index.m3u8" best \
-o /tmp/out.ts \
-f -l debug

....

[stream.hls][debug] Download of segment 73 complete
[stream.hls][debug] Adding segment 95 to queue
[stream.hls][debug] Download of segment 74 complete
[download][out.ts] Written 629.9 MB (1m20s @ 7.2 MB/s)                                                                                                                                                [stream.hls][debug] Adding segment 96 to queue
[stream.hls][debug] Download of segment 75 complete
[stream.hls][debug] Adding segment 97 to queue
[stream.hls][debug] Download of segment 76 complete
[stream.hls][debug] Adding segment 98 to queue
[download][out.ts] Written 658.0 MB (1m21s @ 10.0 MB/s)                                                                                                                                               [stream.hls][debug] Download of segment 77 complete
[stream.hls][debug] Adding segment 99 to queue
[stream.hls][debug] Download of segment 78 complete
[stream.hls][debug] Adding segment 100 to queue
[stream.hls][debug] Download of segment 79 complete
[stream.hls][debug] Adding segment 101 to queue
[download][out.ts] Written 668.3 MB (1m24s @ 6.9 MB/s)                                                                                                                                                [stream.hls][debug] Download of segment 80 complete
[stream.hls][debug] Adding segment 102 to queue
[download][out.ts] Written 696.8 MB (1m25s @ 10.1 MB/s)                                                                                                                                               [stream.hls][debug] Download of segment 81 complete
[stream.hls][debug] Adding segment 103 to queue
<strong>error: Error when reading from stream: Read timeout, exiting</strong>
[stream.hls][debug] Closing worker thread
[stream.hls][debug] Closing writer thread
[stream.hls][debug] Download of segment 82 complete
[cli][info] Stream ended
[cli][info] Closing currently open stream...
</code></pre>

No player involved for me, just downloading to a file... How might one debug this ?

**Comment by hmedia1** [comment]
The `--hls-timeout` documentation is extremely unclear.  It may seem clear in retrospect, but it would be ideal if it actually stated that this is an overriding option that applies to the overall stream and has nothing to do with segments, or retry times - and that if no data is received by the process for TIMEOUT seconds, then the program will **terminate**.
